app/perfunwind.cpp

```diff
@@ -682,6 +691,26 @@ void PerfUnwind::analyze(const PerfRecordSample &sample)
     sendBuffer(buffer);
 }
 
+void PerfUnwind::updateIpCache()
+{
+    const auto& stack = m_currentUnwind.ipStack;
+    for (int i = 0; i < stack.length() - 1; i++) {
+        m_ipcache[m_currentUnwind.sample->pid()][stack[i].address] = {m_currentUnwind.frames, stack[i].frameOffset};
+    }
+}
+
+QVector<int> PerfUnwind::stitchedCallChain()
+{
+    QVector<int> callchain = m_currentUnwind.frames;
+
+    const auto cache = m_ipcache[m_currentUnwind.sample->pid()];
+    if (cache.find(m_currentUnwind.ipStack.last().address) != cache.end()) {
```

Comment:
double lookup, use the result of find() instead! also, the `appendFrames` is going to operate on a temporary copy, so it will never actually change anything? `cache` is const after all?! also once you lookup by address, and once you lookup by pid - that looks very broken?

```
if (m_currentUnwind.ipStack.isEmpty())
    return;

const auto &cache = m_ipCache[m_currentUnwind.sample->pid()];
auto it = cache.find(m_currentUnwind.ipStack.last().address);
if (it != cache.end()) {
    it->appendFramesTo(m_currentUnwind.frames);
}
```

Reply:
I wonder how that even worked